src/client/components/pages/Courses/EnrollmentModal.tsx

```diff
@@ -0,0 +1,321 @@
+
+        // Match any non-alphanumeric characters(also allowing the - character)
+        } else if (new RegExp(/[^\dA-Z-]/gi).test(fieldValue.toString())) {
+          errors.push(`${displayName} cannot contain special characters`);
+
+        // Check that the number > 0
+        } else if (parseInt(fieldValue, 10) < 0) {
+          errors.push(`${displayName} cannot be negative`);
+        }
+        return errors;
+      })
+      .reduce((acc, val) => acc.concat(val), []);
+
+    if (errorMessages.length > 0) {
+      setSaving(false);
+      setValidationErrors(errorMessages);
+    } else {
+      setValidationErrors([]);
+      setSaving(true);
+      const results = await updateCourseInstance(
+        instance.id,
+        {
+          offered: instance.offered,
+          ...sanitizeEnrollmentData(enrollmentData),
+        }
+      );
+      onSave(results);
+      setSaving(false);
+    }
+  };
+
+  /**
+   * Called when the modal is closed. If there are any unsaved changes,
+   * a warning message appears, and the user must confirm discarding the unsaved
+   * changes in order to close the modal. If the user selects cancel, the user
+   * can continue making changes in the modal.
+   */
+  const onModalClose = () => {
+    if (changed) {
+      // eslint-disable-next-line no-alert
+      if (window.confirm('You have unsaved changes. Click \'OK\' to disregard changes, or \'Cancel\' to continue editing.')) {
+        setChanged(false);
+        onClose();
+      }
+    } else {
+      onClose();
+    }
+  };
+
+  return (
+    <Modal
+      ariaLabelledBy="enrollment-modal-header"
+      isVisible={isVisible}
+      closeHandler={onModalClose}
+    >
+      <ModalHeader
+        forwardRef={modalHeaderRef}
+        tabIndex={0}
+      >
+        <span id="enrollment-modal-header">
+          {`Enrollment for ${course?.catalogNumber}, ${currentSemester?.term} ${currentSemester.calendarYear}`}
+        </span>
+      </ModalHeader>
+      <ModalBody>
+        {
+          saving ? (<LoadSpinner>Saving Enrollment Data</LoadSpinner>) : (
+            enrollmentFields.map(({ key, name }) => (
+              <TextInput
+                key={key}
+                id={key}
+                name={key}
+                label={name}
+                value={enrollmentData?.[key]?.toString() || ''}
+                onChange={
+                  ({ target: { value } }: ChangeEvent<HTMLInputElement>) => {
+                    setChanged(true);
+                    setEnrollmentData((state) => ({
+                      ...state,
+                      [key]: value,
+                    }));
+                  }
+                }
+              />
+            ))
+          )
+        }
+        {
+          validationErrors.length > 0
+            ? (
+              <ModalMessage variant={VARIANT.NEGATIVE}>
+                {validationErrors.join(', ')}
+              </ModalMessage>
+            ) : null
+        }
```

Comment:
Our convention for client-side validation has been to show the field-level errors alongside the actual text input (i.e. passing them as the `errorMessage` prop to `TextInput`). This is the case in [Course Admin Modal][1], for instance. By contrast, we've been using the `ModalMessage` for errors returned from the server.

[1]: https://github.com/seas-computing/course-planner/blob/main/src/client/components/pages/Courses/CourseModal.tsx

Reply:
Oh, I see. I did see `ModalMessage` in the course admin modal (which is where I got this idea from) but apparently misunderstood it's purpose a little.